File: src/be/gamelogic/async_gamerunner.py
from .match4game import Match4Game, Match4State, Match4Command, Match4Agent
from copy import deepcopy

class AsyncGamerunner_Match4:
    game: Match4Game
    p1: Match4Agent
    p2: Match4Agent
    cur_player: Match4Agent
    command_histo: list[Match4Command]


    def __init__(self, p1, p2):
        self.game = Match4Game()
        self.p1 = p1
        self.p1.player_id = 1
        self.p2 = p2
        self.p2.player_id = 2
        self.cur_player = p1
        self.command_histo = []

    """
        start a fresh new game and start the playing process
    """
    def start(self):
        self.game.setup()

    def apply_move(self, move: int, player_id: int, is_bot=False):
        if (is_bot):
            command = self.cur_player.take_turn(self.game)
        else:
            command = Match4Command()
            command.column = move - 1
            command.player_id = player_id
        self.game.apply_command(command)
        self.command_histo.append(deepcopy(command))
        self.cur_player = self.p2 if (self.cur_player is self.p1) else self.p1 
        
    def to_dict(self):
        return {
            "state" : self.game.to_dict(),
            "moves": [move.to_dict() for move in self.command_histo] 
        }
    
    def from_dict(self, given: dict):
        self.game = Match4Game.from_dict(given["state"])
        self.command_histo = [Match4Command.from_dict(move) for move in given["moves"]]
        pass
    
    

File: code/fe/src/game_logic/connect4/asyncgamerunner.js
import { Match4Game, Match4Command } from "./match4game.js";

class AsyncGamerunnerMatch4 {
    constructor() {
        this.game = new Match4Game();
        this.p1 = 1
        this.p2 = 2
        this.curPlayer = this.p1;
        this.commandHisto = [];
        this.terminal = false
    }

    /**
     * Start a fresh new game
     * 
     * @returns {(number, number)} - returns the ids for player 1 and player 2
     */
    start() {
        this.game.setup();
        return (this.p1, this.p2)
    }

    /**
     * Apply a move to the game
     * @param {number} move - Column to place the token (1-based index)
     * @param {number} playerId - ID of the player making the move
     */
    applyMove(move, playerId) {
        if (playerId != this.curPlayer) { 
            console.error(" AsynGameRunner: playerId for the given value is bad!!!!")
            return;
        }
        let command = new Match4Command();
        command.column = move - 1; // Convert to 0-based index
        command.playerId = playerId;
        this.game.applyCommand(command);
        this.commandHisto.push(structuredClone(command));
        this.curPlayer = this.curPlayer === this.p1 ? this.p2 : this.p1;
        this.terminal = this.game.isTerminal()
    }

    getBoard() {
        return this.game._state.board
    }

    getTerminal() {
        return this.game._state
    }

    /**
     * Convert game state to a JSON-compatible object
     */
    toDict() {
        return {
            state: this.game.toDict(),
            moves: this.commandHisto.map(move => move.toDict())
        };
    }

    /**
     * Restore game state from a JSON object
     * @param {object} given - The object containing game state and moves
     */
    fromDict(given) {
        this.game = Match4Game.fromDict(given.state);
        this.commandHisto = given.moves.map(move => Match4Command.fromDict(move));
    }
}

// ✅ Export class for use in Node.js
export { AsyncGamerunnerMatch4 };
